guides/source/7_1_release_notes.md

```diff
@@ -150,7 +150,51 @@ and reporting of the bulk enqueuing process.
 
 ### Composite primary keys
 
-TODO: Add description
+Applications can now create database tables with composite primary keys, and Rails will be able to derive these keys directly from the schema. This feature is particularly beneficial for many-to-many relationships and other complex data models where a single column is insufficient to uniquely identify a record.
+
+Active Record has been updated to fully support composite primary keys. All parts of the composite primary key will be included in the SQL generated by the `#reload`, `#update`, and `#delete methods`. Methods like `#first` and `#last` will use the full composite primary key in the `ORDER BY` statement.
+
+The `query_constraints`` macro can be used as a "virtual primary key" to achieve the same behaviors without modifying the database schema.
+
+Similarly, the `query_constraints:` option is accepted by associations to serve as a composite foreign key. This option configures the list of columns used for accessing the associated record.
```

Comment:
Tiny nit: I'd use active instead of passive voice here to make this more direct
```suggestion
Similarly, associations accept a `query_constraints:` option. This option serves as a composite foreign key, configuring the list of columns used for accessing the associated record.
```